Warn users if PlantUML diagrams are found but Graphviz is not installed

tlylt opened the pull request:
**What is the purpose of this pull request?**

- [ ] Documentation update
- [x] Bug fix
- [ ] Feature addition or enhancement
- [ ] Code maintenance
- [ ] Others, please explain:

<!--
  If this pull request is addressing an issue, link to the issue: "Fixes #xxx" or "Resolves #xxx"

  If the pull request completely addresses the issue, use one of the issue closing keywords. https://docs.github.com/en/enterprise/2.16/user/github/managing-your-work-on-github/closing-issues-using-keywords

  Otherwise, elaborate further on the rationale of this pull request as needed
-->

**Overview of changes:**
Related to #1903
- Add a check to verify that Graphviz is installed before processing PlantUML diagrams.
- This will be done once per build/refresh (in the case of markbind serve)
- A warning will be logged to the console
	- ![image](https://user-images.githubusercontent.com/41845017/163904240-eb71f2d0-638e-4608-bf3f-be12235687e7.png)
- I am leaving the decision and the work to fail hard (or not) to development in #1863, as that will involve changing existing test cases and how we decide what kinds of errors to totally stop the build. This PR simply logs the warning but does not error out.
- Include a way to disable the warning:
	- https://deploy-preview-1912--markbind-master.netlify.app/userguide/sitejsonfile#plantumlcheck
	- ![image](https://user-images.githubusercontent.com/41845017/166176946-f03aa13b-4a49-4f4e-95f3-1d6f96eb46dc.png)

- Lastly, make the warning in our UG even more explicit:
	- https://deploy-preview-1912--markbind-master.netlify.app/userguide/components/imagesanddiagrams#diagrams
	- ![image](https://user-images.githubusercontent.com/41845017/166176881-741cc1bb-6682-48fa-8fb6-87e875484792.png)



**Anything you'd like to highlight / discuss:**
nil

**Testing instructions:**
- Checkout to this PR
- insert the following (or any other puml code) in a .md file:
```md
<puml width="300">
@startuml
alice -> bob ++ : hello
bob -> bob ++ : self call
bob -> bib ++  #005500 : hello
bob -> george ** : create
return done
return rc
bob -> george !! : delete
return success
@enduml
</puml>
```
- Run `markbind build` or `markbind serve`, if Graphviz is installed, no warnings should be found, else, a warning should be found (exactly once even if u have multiple PlantUML diagrams)

<!--
  Any special testing instructions **not including** our automated tests.
-->

**Proposed commit message: (wrap lines at 72 characters)**
Add Graphviz check when PlantUML diagrams are used

Without Graphviz, PlantUML diagrams generated may
contain errors. Users may not be aware that diagrams are
generated incorrectly if they do not serve up the pages.

Let's add a check and warn users in the console if the
installation of Graphviz is missing.

This helps to keep users informed in the console when
the needed installation is missing. Users are also informed
of a way to disable the warning in `site.json`.

<!--
  See this link for more info on how to write a good commit message:
  https://oss-generic.github.io/process/docs/FormatsAndConventions.html#commit-message

  As best as possible, write a succinct commit title in 50 characters
-->

---

**Checklist:** :ballot_box_with_check:

<!-- Leave non-applicable items unchecked -->

- [x] Updated the documentation for feature additions and enhancements
- [ ] Added tests for bug fixes or features
- [x] Linked all related issues
- [x] No unrelated changes <!-- Its tempting, but increases the reviewer's work, and really pollutes the commit history =( -->

<!--
  We'll try our best to get to your PR within a week.
  If we haven't gotten to it then, or if your pull request resolves an urgent item, feel free to give us a ping!
-->

damithc commented:
This is a good feature to add @tlylt 

Perhaps the CI info is a better info for a minimized panel rather than a popover, as it contains substantial amount of info, and stuff the user might want to copy-paste? Alternatively, it can simply link to another page that has the full CI details.

As for the error message, should we mention that only some (or which) puml diagrams require Graphviz? I'm thinking of the situation in which the site has puml diagrams that don't actually need Graphviz and yet this warning keeps showing up every time. I suppose making the error message smart enough to figure out if Graphviz is actually needed is not feasible or worth the effort?

tlylt commented:
> Perhaps the CI info is a better info for a minimized panel rather than a popover, as it contains substantial amount of info, and stuff the user might want to copy-paste? Alternatively, it can simply link to another page that has the full CI details.

Thanks for the suggestion, I have updated the docs to use a panel and included more details. (see PR description). Is this ok?

> As for the error message, should we mention that only some (or which) puml diagrams require Graphviz? I'm thinking of the situation in which the site has puml diagrams that don't actually need Graphviz and yet this warning keeps showing up every time. I suppose making the error message smart enough to figure out if Graphviz is actually needed is not feasible or worth the effort?

I have updated the docs to mention what's stated in the PlantUML website (that sequence and activity diagram work without Graphviz). For the warning, as I am doing this check before processing any of the puml diagrams, this implementation would not be able to identify the diagram type. This warning is therefore more like our broken link reporting. Do you think adding something like "Ignore this warning if it was intended." will be better? 

Ideally, we should have the warning be smarter, which could be done with some adjustments to how we process the PlantUML diagrams. I have something in mind to try out but because of the other considerations involved (e.g. performance, adjustments needed to the test cases), might take some time before it's done. Hence I am also ok if we want to remove this warning log in this PR and keep the documentation updated for now.

damithc commented:
> Thanks for the suggestion, I have updated the docs to use a panel and included more details. (see PR description). Is this ok?

Yes, this looks better. A slight worry is that a sizeable chunk of the yml file is duplicated here (I expect it is also given somewhere else in our docs?). Can go out of sync if we update the CI documentation later. 
Also, not strictly a 'tip', right? More like an 'example' of how to configure a CI, especially because we don't explain how to do it in other CI environments?


> Ideally, we should have the warning be smarter, which could be done with some adjustments to how we process the PlantUML diagrams. I have something in mind to try out but because of the other considerations involved (e.g. performance, adjustments needed to the test cases), might take some time before it's done. Hence I am also ok if we want to remove this warning log in this PR and keep the documentation updated for now.

Yes, we certainly don't want to lose performance over this. We _can_ consider a flag e.g., `requireGraphviz` in site.json to switch off this warning, although not sure if that would be going too far.

tlylt commented:
> Yes, this looks better. A slight worry is that a sizeable chunk of the yml file is duplicated here (I expect it is also given somewhere else in our docs?). Can go out of sync if we update the CI documentation later. Also, not strictly a 'tip', right? More like an 'example' of how to configure a CI, especially because we don't explain how to do it in other CI environments?

Updated to mention only the required steps.
![image](https://user-images.githubusercontent.com/41845017/163926630-4b0b499a-d75a-4237-9153-5db062b8f6df.png)

> Yes, we certainly don't want to lose performance over this. We _can_ consider a flag e.g., `requireGraphviz` in site.json to switch off this warning, although not sure if that would be going too far.

Hmm in that case I will leave this PR here for the time being.

damithc commented:
> Hmm in that case I will leave this PR here for the time being.

If the flag is easy to add, go ahead and add it. I don't think we want a case where users see a warning that technically should not be there (because their PUML diagrams don't actually need Graphviz) and have no way to get rid of it. We can remove the flag when (and if) we make the error message 'smarter'? What do you think?

tlylt commented:
> > Hmm in that case I will leave this PR here for the time being.
> 
> If the flag is easy to add, go ahead and add it. I don't think we want a case where users see a warning that technically should not be there (because their PUML diagrams don't actually need Graphviz) and have no way to get rid of it. We can remove the flag when (and if) we make the error message 'smarter'? What do you think?

Sure I will add the option to suppress the warning.

jonahtanjz commented on packages/core/src/Site/index.js:
```diff
@@ -950,7 +951,8 @@ class Site {
         || !_.isEqual(oldSiteConfig.timeZone, this.siteConfig.timeZone)
         || !_.isEqual(oldSiteConfig.locale, this.siteConfig.locale)
-        || !_.isEqual(oldSiteConfig.intrasiteLinkValidation, this.siteConfig.intrasiteLinkValidation);
+      || !_.isEqual(oldSiteConfig.intrasiteLinkValidation, this.siteConfig.intrasiteLinkValidation)
+      || !_.isEqual(oldSiteConfig.plantumlCheck, this.siteConfig.plantumlCheck);
```
I think these 2 lines should be indented by 2 more spaces?

jonahtanjz commented on packages/core/src/plugins/default/markbind-plugin-plantuml.js:
```diff
@@ -84,12 +86,23 @@ module.exports = {
       return;
     }
+    if (config.plantumlCheck && !graphvizCheckCompleted) {
+      exec('dot -V', (error, stdout, stderr) => {
```
```suggestion
      exec(`java -jar "${JAR_PATH}" -testdot`, (error, stdout, stderr) => {
```

Would this be better since it is the recommended testing method on the [plantuml site](https://plantuml.com/graphviz-dot#fa6a5a3224d7da66)? The original should work as well for most cases (except when Graphviz is installed in a custom path).

jonahtanjz reviewed (commented):
Thanks @tlylt

Just 2 comments, the rest looks good :)
